File: nrtest3/code/basics/06/spec/models/project_spec.rb
#---
# Excerpted from "Rails 5 Test Prescriptions",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material,
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose.
# Visit http://www.pragmaticprogrammer.com/titles/nrtest3 for more book information.
#---
require "rails_helper"

RSpec.describe Project do

  describe "completion" do
    let(:project) { Project.new }
    let(:task) { Task.new }

    it "considers a project with no tasks to be done" do
      expect(project).to be_done
    end

    it "knows that a project with an incomplete task is not done" do
      project.tasks << task
      expect(project).not_to be_done
    end

    it "marks a project done if its tasks are done" do
      project.tasks << task
      task.mark_completed
      expect(project).to be_done
    end
  end

  #
  describe "estimates" do
    let(:project) { Project.new }
    let(:done) { Task.new(size: 2, completed: true) }
    let(:small_not_done) { Task.new(size: 1) }
    let(:large_not_done) { Task.new(size: 4) }

    before(:example) do
      project.tasks = [done, small_not_done, large_not_done]
    end

    it "can calculate total size" do
      expect(project.total_size).to eq(7)
    end

    it "can calculate remaining size" do
      expect(project.remaining_size).to eq(5)
    end
  end
  #

end

File: nrtest3/code/security/03/spec/javascripts/project_spec.js
/***
 * Excerpted from "Rails 5 Test Prescriptions",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/nrtest3 for more book information.
***/
// #
import {Project} from "../../app/javascript/packs/project.js"
import {Task} from "../../app/javascript/packs/task.js"
import {TaskUpdater} from "../../app/javascript/packs/task_updater.js"
import td from "testdouble/dist/testdouble"
import tdJasmine from "testdouble-jasmine"
tdJasmine.use(td)
// #

describe("Projects", () => {
  let project

  beforeEach(() => {
    project = new Project(1)
    project.appendTask(new Task("Start Project", 1))
    project.appendTask(new Task("Middle Project", 2))
    project.appendTask(new Task("End Project", 3))
  })

  it("can identify the first element of a project", () => {
    expect(project.firstTask().name).toEqual("Start Project")
    expect(project.firstTask().isFirst()).toBeTruthy()
    expect(project.firstTask().isLast()).toBeFalsy()
    expect(project.firstTask().index()).toEqual(0)
  })

  it("can identify the last element of a project", () => {
    expect(project.lastTask().name).toEqual("End Project")
    expect(project.lastTask().isLast()).toBeTruthy()
    expect(project.lastTask().isFirst()).toBeFalsy()
    expect(project.lastTask().index()).toEqual(2)
  })

  // #
  it("can move a task up", () => {
    const FakeUpdater = td.constructor(TaskUpdater)
    project.tasks[1].updater = new FakeUpdater()
    project.tasks[1].moveUp()
    expect(project.firstTask().name).toEqual("Middle Project")
    expect(project.tasks[1].name).toEqual("Start Project")
    expect(project.lastTask().name).toEqual("End Project")
    td.verify(FakeUpdater.prototype.update("up"))
  })

  it("can move a task down", () => {
    const FakeUpdater = td.constructor(TaskUpdater)
    project.tasks[1].updater = new FakeUpdater()
    project.tasks[1].moveDown()
    expect(project.firstTask().name).toEqual("Start Project")
    expect(project.tasks[1].name).toEqual("End Project")
    expect(project.lastTask().name).toEqual("Middle Project")
    expect().toVerify(FakeUpdater.prototype.update("down"))
  })
  // #

  it("handles asking for the top task to move up", () => {
    project.firstTask().moveUp()
    expect(project.firstTask().name).toEqual("Start Project")
    expect(project.tasks[1].name).toEqual("Middle Project")
    expect(project.lastTask().name).toEqual("End Project")
  })

  it("handles asking for the bottom task to move up", () => {
    project.lastTask().moveDown()
    expect(project.firstTask().name).toEqual("Start Project")
    expect(project.tasks[1].name).toEqual("Middle Project")
    expect(project.lastTask().name).toEqual("End Project")
  })
})
